src/pylibs/db/sql_adapter.py

```diff
@@ -0,0 +1,158 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from sqlalchemy import create_engine, MetaData, Table, Column, String, Date, LargeBinary, Boolean
+import bcrypt
+import logging
+import random
+from datetime import datetime
+
+# timestamp format for birthdays
+TIMESTAMP_FORMAT = '%Y-%m-%d'
+
+
+class SqlAdapter:
+    def __init__(self, uri):
+        # connect to db
+        self.engine = create_engine(uri)
+        self.users_table = Table(
+            'users',
+            MetaData(self.engine),
+            Column('accountid', String),
+            Column('username', String),
+            Column('passhash', LargeBinary),
+            Column('firstname', String),
+            Column('lastname', String),
+            Column('birthday', Date),
+            Column('timezone', String),
+            Column('address', String),
+            Column('state', String),
+            Column('zip', String),
+            Column('ssn', String),
+        )
+        self.contacts_table = Table(
+            'contacts',
+            MetaData(self.engine),
+            Column('username', String),
+            Column('label', String),
+            Column('account_num', String),
+            Column('routing_num', String),
+            Column('is_external', Boolean),
+        )
+        # open a connection
+        self.conn = self.engine.connect()
```

Comment:
@djmailhot @DanSanche not sure if opening a connection and not closing it until end of pod lifecycle is a best practice. I dont have alot of SQL experience, but from what I found online it seems we should be atomically opening and closing connections per operation.

Reply:
That's a good point!

SQL has limits to the number of concurrent connections (PostgreSQL defaults to 100), so opening/closing connections on each request would 1) limit the maximum QPS of the service, and 2) add overhead latency to each request.

HOWEVER, it appears that sqlalchemy automatically employs [connection pooling](https://docs.sqlalchemy.org/en/13/core/pooling.html), which means **we can and should indeed open/close connections on each request**.

Reply:
aha cool, I will make that change

Reply:
Done